### PR TITLE
optimize cross entropy kernel by using reduce.

### PR DESCRIPTION
**Pull request by lcy-seso**
fixes https://github.com/PaddlePaddle/Paddle/issues/4236

- [x] optimize the forward kernel with soft labels
- [x] optimize the backward kernel with soft labels
- [x]  Cpu forward kernel by directly calling Eigen.
- [x]  Cpu backward kernel by directly calling Eigen.

### CODE REVIEW

File: python/paddle/v2/framework/tests/test_cross_entropy_op.py [hunk 1]
```diff
@@ -74,15 +75,15 @@ def setUp(self):
             dtype="float32")
         cross_entropy2 = (-label * np.log(X)).sum(
             axis=1, keepdims=True).astype("float32")
-        self.inputs = {'X': X, 'Label': label}
-        self.outputs = {'Y': cross_entropy}
-        self.attrs = {'soft_label': 1}
+        self.inputs = {"X": X, "Label": label}
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply python/paddle/v2/framework/tests/test_cross_entropy_op.py]: Yes, I know. But I prefer to keep consistent in one file.

File: paddle/operators/cross_entropy_op.cu [hunk 2]
```diff
@@ -138,6 +155,11 @@ class CrossEntropyGradientOpCUDAKernel : public framework::OpKernel {
     // TODO(qingqing): launch kernel on specified stream
     // base on ExecutionContext.
     if (ctx.Attr<int>("soft_label") == 1) {
+      int block_x = 32;
+      int block_y = 32;
+      dim3 block(block_x, block_y);
+      dim3 grid((n + block_x - 1) / block_x, (d + block_y - 1) / block_y);
```
Comment (qingqing01) [inline paddle/operators/cross_entropy_op.cu]: 这样分配二维block，相比一维的block空的线程更多，这个计算实际一维block就够了, 还是使用非soft-label的方式：

```cpp
int block = 512;
int grid = (n * d + block - 1) / block;
```

kernel 改成：

```cpp
for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < N * D;
    i += blockDim.x * gridDim.x) {
  int id_dy = ids / D; 
  dX[i] = -label[i] * dY[id_dy] / X[i];
}
```

另外如果方便的话，可以帮忙一起修复下:

> // TODO(qingqing): launch kernel on specified stream
> // base on ExecutionContext.

谢谢了，当时写的时候ExecutionContext里还没有stream.

File: paddle/operators/cross_entropy_op.cu [hunk 3]
```diff
@@ -104,8 +120,9 @@ class CrossEntropyOpCUDAKernel : public framework::OpKernel {
     // base on ExecutionContext.
     if (ctx.Attr<int>("soft_label") == 1) {
       auto* label_data = ctx.Input<Tensor>("Label")->data<T>();
-      SoftCrossEntropyKernel<T><<<grid, block>>>(y_data, x_data, label_data, n,
-                                                 d);
+      grid = d;
+      SoftCrossEntropyKernel<T, 512><<<grid, block>>>(y_data, x_data,
```
Comment (qingqing01) [inline paddle/operators/cross_entropy_op.cu]: 去掉line 123， 写成：

```cpp
SoftCrossEntropyKernel<T, 512><<<d, block>>>
```

Comment (lcy-seso) [reply paddle/operators/cross_entropy_op.cu]: done.

File: paddle/operators/cross_entropy_op.cu (hunk 2 above)
Comment (lcy-seso) [reply paddle/operators/cross_entropy_op.cu]: done.

**Comment by lcy-seso** [comment]
Not finished yet, please do not review. Thank you.

**Comment by lcy-seso** [comment]
all finished, please review this, thanks, everyone. @qingqing01 @Xreki

File: paddle/operators/cross_entropy_op.cc [hunk 4]
```diff
@@ -98,24 +102,38 @@ class CrossEntropyOpMaker : public framework::OpProtoAndCheckerMaker {
...
-    AddAttr<bool>("soft_label", "Is soft label. Default zero.")
+    AddInput("X",
+             "(Tensor, default Tensor<float>), a 2-D tensor with shape N x D, "
+             "where N is the batch size and D is the number of classes. "
+             "This input is a probability computed by the previous operator, "
+             "which is almost always the result of a softmax operator.");
+    AddInput("Label",
+             "(Tensor, default Tensor<int>), the ground truth which is "
+             "a 1-D or 2-D tensor. "
+             "When softLabel is set to 0, `Label` is a Tensor<int> with shape "
```
Comment (qingqing01) [inline paddle/operators/cross_entropy_op.cc]: set to 0 -> false.

File: paddle/operators/cross_entropy_op.cc [hunk 5]
```diff
@@ -98,24 +102,38 @@ class CrossEntropyOpMaker : public framework::OpProtoAndCheckerMaker {
...
+             "[N x 1]. "
+             "When softLabel is set to 1, `Label` is a Tensor<float/double> "
```
Comment (qingqing01) [inline paddle/operators/cross_entropy_op.cc]: set to 1 -> set to true.

File: paddle/operators/cross_entropy_op.cc (hunk 4 above)
Comment (lcy-seso) [reply paddle/operators/cross_entropy_op.cc]: done

File: paddle/operators/cross_entropy_op.cc (hunk 5 above)
Comment (lcy-seso) [reply paddle/operators/cross_entropy_op.cc]: done
